subsys/shell/shell_mqtt.c

```diff
@@ -0,0 +1,775 @@
+
+	rc = mqtt_publish(&sh_mqtt->client_ctx, &sh_mqtt->pub_data);
+	memset(&sh_mqtt->line_out, 0, sizeof(sh_mqtt->line_out));
+	if (rc != 0) {
+		LOG_ERR("MQTT publish error: %d", rc);
+		return rc;
+	}
+
+	/* Arbitrary delay to not kill the session */
+	if (!is_work) {
+		k_msleep(MQTT_SEND_DELAY_MS);
+	}
+
+	return rc;
+}
+
+static void shell_mqtt_send_work(struct k_work *work)
+{
+	struct shell_mqtt *ctx;
+
+	ctx = CONTAINER_OF(work, struct shell_mqtt, mqtt_work);
+
+	if (shell_mqtt_send(true) != 0) {
+		ctx.state = SHELL_MQTT_OP_ERROR;
+		shell_mqtt_handle_error(ctx);
+	}
+}
+
+/*
+ * Network connection event handler
+ */
+static void event_handler(struct net_mgmt_event_callback *cb, uint32_t mgmt_event,
+			  struct net_if *iface)
+{
+	if ((mgmt_event & EVENT_MASK) != mgmt_event) {
+		return;
+	}
+
+	if (mgmt_event == NET_EVENT_L4_CONNECTED) {
+		if (sh_mqtt->state == SHELL_MQTT_OP_WAIT_NET ||
+		    sh_mqtt->state == SHELL_MQTT_OP_ERROR) {
+			LOG_DBG("Network connected, connecting to MQTT broker");
+			LOG_DBG("Initializing shell MQTT broker & client");
+			broker_init(sh_mqtt);
+			client_init(sh_mqtt);
+			k_work_init_delayable(&sh_mqtt->mqtt_work, shell_mqtt_connect_work);
+			(void)k_work_reschedule_for_queue(&sh_mqtt->sh_mqtt_workq,
+							  &sh_mqtt->mqtt_work, K_SECONDS(2));
+			sh_mqtt->state = SHELL_MQTT_OP_CONNECTING;
+		}
+	} else if (mgmt_event == NET_EVENT_L4_DISCONNECTED) {
+		LOG_DBG("No network connection");
+		(void)k_work_cancel_delayable_sync(&sh_mqtt->mqtt_work, &sh_mqtt->work_sync);
+		(void)k_work_cancel_delayable_sync(&sh_mqtt->send_work, &sh_mqtt->work_sync);
+		(void)mqtt_abort(&sh_mqtt->client_ctx);
+		clear_fds(sh_mqtt);
+
+		sh_mqtt->state = SHELL_MQTT_OP_WAIT_NET;
+	}
+}
+
+static void mqtt_evt_handler(struct mqtt_client *const client, const struct mqtt_evt *evt)
+{
+	switch (evt->type) {
+	case MQTT_EVT_SUBACK:
+		LOG_DBG("SUBACK packet id: %u", evt->param.suback.message_id);
+		sh_mqtt->state = SHELL_MQTT_OP_SUBSCRIBED;
+		break;
+
+	case MQTT_EVT_UNSUBACK:
+		LOG_DBG("UNSUBACK packet id: %u", evt->param.suback.message_id);
+		sh_mqtt->state = SHELL_MQTT_OP_CONNECTED;
+		(void)k_work_cancel_delayable_sync(&sh_mqtt->mqtt_work, &sh_mqtt->work_sync);
+		k_work_init_delayable(&sh_mqtt->mqtt_work, shell_mqtt_subscribe_work);
+		(void)k_work_reschedule_for_queue(&sh_mqtt->sh_mqtt_workq, &sh_mqtt->mqtt_work,
+						  K_SECONDS(MQTT_PROCESS_INTERVAL));
+		break;
+
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT connect failed %d", evt->result);
+			break;
+		}
+
+		sh_mqtt->state = SHELL_MQTT_OP_CONNECTED;
+		LOG_DBG("MQTT client connected!");
+
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		if (sh_mqtt->state == SHELL_MQTT_OP_WAIT_NET) {
+			/* MQTT_EVT_DISCONNECT likely wont happen while the shell mqtt is in
+			 * SHELL_MQTT_OP_WAIT_NET state, because mqtt_abort is invoked when the
+			 * network is lost (NET_EVENT_L4_DISCONNECTED), and that will not generate
+			 * MQTT_EVT_DISCONNECT event.
```

Comment:
```c
			    ... because mqtt_abort is invoked when the
			 * network is lost (NET_EVENT_L4_DISCONNECTED), and that will not generate
			 * MQTT_EVT_DISCONNECT event.
 ```

This isn't true.